app/models/cart.rb

```diff
@@ -43,4 +43,98 @@ def subtotal_of(item_id)
   def limit_reached?(item_id)
     count_of(item_id) == Item.find(item_id).inventory
   end
+
+  def cart_merchants
+    items.map do |item|
+      item.merchant_id
+    end.uniq
+  end
+
+  def valid_discounts
+    possible_discounts = cart_merchants.map do |merchant_id|
+      Merchant.find(merchant_id).discounts
+    end.flatten
+    discount_qualifications_met(possible_discounts)
+  end
+
+  def item_on_sale(item, discount)
+    !DiscountItem.where('item_id = ? AND discount_id = ?', item.id, discount.id).empty? && item.merchant_id == discount.merchant_id
+  end
+
+  def validate_attributes(possible_discounts)
+    totals = Hash.new { |hash, key| hash[key] = {quantity: 0, value: 0} }
+    possible_discounts.each do |discount|
+      items.each do |item|
```

Comment:
I have this nested iterations with item in multiple places and I would love to replace it with something that a) reduces nested iteration and b) reduces database calls (items is a cart method here that calls the database every time).